stream_batch.go

```diff
@@ -0,0 +1,220 @@
+package scylla_cdc
+
+import (
+	"context"
+	"encoding/base64"
+	"sync/atomic"
+	"time"
+
+	"github.com/gocql/gocql"
+)
+
+type streamBatchReader struct {
+	config         *ReaderConfig
+	generationTime time.Time
+	streams        []StreamID
+	keyspaceName   string
+	tableName      string
+
+	lastTimestamp gocql.UUID
+	endTimestamp  atomic.Value
+
+	interruptCh chan struct{}
+}
+
+func newStreamBatchReader(
+	config *ReaderConfig,
+	generationTime time.Time,
+	streams []StreamID,
+	keyspaceName string,
+	tableName string,
+	startFrom gocql.UUID,
+) *streamBatchReader {
+	return &streamBatchReader{
+		config:         config,
+		generationTime: generationTime,
+		streams:        streams,
+		keyspaceName:   keyspaceName,
+		tableName:      tableName,
+
+		lastTimestamp: startFrom,
+
+		interruptCh: make(chan struct{}, 1),
+	}
+}
+
+func (sbr *streamBatchReader) run(ctx context.Context) error {
+	baseTableName := sbr.keyspaceName + "." + sbr.tableName
+
+	oldProgress := make(map[string]gocql.UUID, len(sbr.streams))
+	for _, stream := range sbr.streams {
+		progress, err := sbr.config.ProgressManager.GetProgress(sbr.generationTime, baseTableName, stream)
+		if err != nil {
+			return err
+		}
+		oldProgress[string(stream)] = progress.Time
+	}
+
+	minTime := oldProgress[string(sbr.streams[0])]
```

Comment:
please explain in a comment what is happening here, i.e. why we want to read from the min progress

Reply:
Done